File: second course/PDOO/civitasJava/src/civitas/Tablero.java

package civitas;

import java.util.ArrayList;

class Tablero{
  
  private int numCasillaCarcel;        
  private int porSalida;    
  private Boolean tieneJuez;    
	
	//Atributos de referencia
	private ArrayList <Casilla> casillas;  

	
  public Tablero(int posCarcel){ //TESTED
    if(posCarcel >= 1)
      numCasillaCarcel = posCarcel;
	
    else
      numCasillaCarcel = 1;
       
    casillas = new ArrayList<>();
    casillas.add(new Casilla("Salida"));
        
    porSalida = 0;
        
	  tieneJuez = false;
  }
    
	private Boolean correcto(){ //TESTED
    return (casillas.size() > numCasillaCarcel && tieneJuez);
  }

  private Boolean correcto(int numCasilla){ //TESTED
    Boolean indiceValido = (numCasilla >= 0 && numCasilla < casillas.size());
        
    return(correcto() && indiceValido);
  }

  int getCarcel(){ //TESTED
    return numCasillaCarcel;
  }
    
  int getPorSalida(){ //TESTED
	  int vecesPorSalida = porSalida;
	
	  if(porSalida > 0)
			--porSalida;
		  
    return vecesPorSalida; 
  }
    
  void añadeCasilla(Casilla casilla){ //TESTED
    if(casillas.size() == numCasillaCarcel)//Precomprobacion
      casillas.add(new Casilla("Carcel"));
        
    casillas.add(casilla);
        
    if(casillas.size() == numCasillaCarcel)//Poscomprobacion
      casillas.add(new Casilla("Carcel"));
  }
    
  void añadeJuez(){ //TESTED
    if(!tieneJuez){
      añadeCasilla(new Casilla(numCasillaCarcel, "Juez"));
			tieneJuez = true;
    }   
  }
    
	public int getTamTablero(){ //TESTED
	  return casillas.size();
	}
	
  public Casilla getCasilla(int numCasilla){ //TESTED
    return correcto(numCasilla)? casillas.get(numCasilla) : null;
  }
    
  int nuevaPosicion(int actual, int tirada){
	  int nuevaPosActual = -1;//preinicializacion a tablero no correcto
	
	  if(correcto()){//Comprobamos si el tablero es correcto
									//Calculamos nueva posicion
			nuevaPosActual = (actual + tirada) % casillas.size();
			
			if(actual + tirada != nuevaPosActual)//Paso por salida?
		  ++porSalida;
	  }
	
	  return nuevaPosActual;
	}
	
	int calcularTirada(int origen, int destino){ //TESTED
	  int tirada = destino - origen;
		
	  if(tirada < 0)//Ajuste por si paso por salida
			tirada += casillas.size();
		
	  return tirada;
	}
    
	public static void main(String [] args){
		
		Tablero t = new Tablero(3);
		
		System.out.println("FIN");
		System.out.println("FIN");
	}
}

File: second course/PDOO/civitasRuby/pkg/civitasRuby-0.0.1/lib/tablero.rb
# encoding:utf-8

require_relative "casilla"

module Civitas
	
	class Tablero
		
		attr_reader :num_casilla_carcel
		
		
		def initialize(pos_carcel)
			if(pos_carcel >= 1)
				@num_casilla_carcel = pos_carcel
			else
				@num_casilla_carcel = 1
			end
			
			@casillas = Array.new()
			@casillas << Casilla.new_por_nombre("Salida")
			
			@por_salida = 0
			
			@tiene_juez = false			
		end
		
		
		private
		
		def correct()
			return (@casillas.size() > @num_casilla_carcel) && @tiene_juez
		end
		
	
		def correcto(num_casilla)
			indice_valido = (num_casilla >= 0 && num_casilla < @casillas.size()) 
			
			return correct() && indice_valido 
		end
		
		
		public
		
		def por_salida()
			veces_por_salida = @por_salida
			
			if(@por_salida > 0)
				@por_salida -= 1
			end
			
			return veces_por_salida
		end
		
		def tam_tablero()
			return @casillas.size()
		end
		
		
		def añade_casilla(casilla)
			if(@casillas.size() == @num_casilla_carcel)
				@casillas << Casilla.new_por_nombre("Carcel")
			end
			
			@casillas << casilla
			
			if(@casillas.size() == @num_casilla_carcel)
				@casillas << Casilla.new_por_nombre("Carcel")
			end
			
		end
		
	
		def añade_juez()
			if(!@tiene_juez)
				añade_casilla(Casilla.new_por_juez(@num_casilla_carcel, "Juez"))
				@tiene_juez = true
			end
		end
		
		
		def casilla(num_casilla)
			return correcto(num_casilla)? @casillas[num_casilla] : nil
		end
		
		
		def nueva_posicion(actual, tirada)
			nueva_pos_actual = -1
			
			if(correct())
				nueva_pos_actual = (actual + tirada) % @casillas.size()
				
				if(actual + tirada != nueva_pos_actual)
					@por_salida += 1
				end
			end
			
			return nueva_pos_actual
		end
		
		
		def calcular_tirada(origen, destino)
			tirada = destino - origen
			
			if(tirada < 0)
				tirada += @casillas.size()
			end
			
			return tirada	
		end
		
		
	end   
end
